flatsurf/geometry/hyperbolic.py

```diff
@@ -0,0 +1,774 @@
+        base_ring = base_ring or QQ
+
+        from sage.categories.all import Sets
+        category = category or Sets()
+
+        return super(HyperbolicPlane, cls).__classcall__(cls, base_ring=base_ring, category=category)
+
+    def __init__(self, base_ring, category):
+        r"""
+        Create the hyperbolic plane over ``base_ring``.
+
+        TESTS::
+
+            sage: from flatsurf.geometry.hyperbolic import HyperbolicPlane
+
+            sage: TestSuite(HyperbolicPlane(QQ)).run()
+            sage: TestSuite(HyperbolicPlane(AA)).run()
+
+        """
+        super().__init__(category=category)
+        self._base_ring = base_ring
+
+    def _an_element_(self):
+        r"""
+        Return an element of the hyperbolic plane (mostly for testing.)
+
+        EXAMPLES::
+
+            sage: from flatsurf.geometry.hyperbolic import HyperbolicPlane
+
+            sage: HyperbolicPlane().an_element()
+            0
+
+        """
+        return self.real(0)
+
+    def some_elements(self):
+        r"""
+        Return some representative convex subsets for automated testing.
+
+        EXAMPLES::
+
+            sage: from flatsurf.geometry.hyperbolic import HyperbolicPlane
+
+            sage: HyperbolicPlane().some_elements()
+            [{}, ∞, 0, 1, -1]
+
+        """
+        # TODO: Return more elements.
+        return [self.empty_set(), self.infinity(), self.real(0), self.real(1), self.real(-1)]
+
+    def _element_constructor_(self, x):
+        r"""
+        Return ``x`` as an element of the hyperbolic plane.
+
+        EXAMPLES::
+
+            sage: from flatsurf.geometry.hyperbolic import HyperbolicPlane
+
+            sage: H = HyperbolicPlane(QQ)
+            sage: H(H.an_element()) in H
+            True
+
+        Base ring elements can be converted to ideal points::
+
+            sage: H(1)
+            1
+
+        The point at infinity in the half plane model can be written directly::
+
+            sage: H(oo)
+            ∞
+
+        Elements can be converted between hyperbolic planes with compatible base rings::
+
+            sage: HyperbolicPlane(AA)(H(1))
+            1
+
+        """
+        if x.parent() is self:
+            return x
+
+        from sage.all import Infinity
+        if x is Infinity:
+            return self.infinity()
+
+        if x in self.base_ring():
+            return self.real(x)
+
+        if isinstance(x, HyperbolicConvexSubset):
+            return x.change_ring(self.base_ring())
+
+        raise NotImplementedError("Cannot create a subset of the hyperbolic plane from this element yet.")
+
+    def base_ring(self):
+        r"""
+        Return the base ring over which objects in the plane are defined.
+
+        EXAMPLES::
+
+            sage: from flatsurf.geometry.hyperbolic import HyperbolicPlane
+
+            sage: HyperbolicPlane().base_ring()
+            Rational Field
+
+        """
+        return self._base_ring
+
+    def infinity(self):
```

Comment:
We have a very generic looking `class HyperbolicPlane` but very model specific name such as `infinity` or `chord`. I find it a bit confusing as a user interface. Couldn't we use something like `HyperbolicPlane.PoincareHalfPlane.infinity`?

Reply:
I considered this option. In the end, `HyperbolicPlane` is just a factory that builds things, so I wanted it to be convenient and not hide things away like this. I tried to find names that leave no space for confusion but maybe that's not the case.

Reply:
That is true that `infinity` only make sense in the Poincaré upper half plane.

Reply:
From a user interface point of view, I think I like better the sage way (I changed the names to make them look better)
```
HyperbolicPlane().upper_half_plane().point(0)
HyperbolicPlane().poincare_disc().point(1/2)
HyperbolicPlane().klein_disc().point(0)
HyperbolicPlane().hyperboloid().point(0, 0, 1)
```
Though the sage internal is not reasonable to me (one coordinate system implementation per model).

Reply:
I prefer the model parameter, i.e., `point(whatever, model="klein")`. Let's discuss this tomorrow.